alibi_detect/utils/pytorch/kernels.py

```diff
@@ -20,21 +56,307 @@ def sigma_median(x: torch.Tensor, y: torch.Tensor, dist: torch.Tensor) -> torch.
 
     Returns
     -------
-    The computed bandwidth, `sigma`.
+    The logrithm of the computed bandwidth, `log-sigma`.
     """
     n = min(x.shape[0], y.shape[0])
     n = n if (x[:n] == y[:n]).all() and x.shape == y.shape else 0
     n_median = n + (np.prod(dist.shape) - n) // 2 - 1
     sigma = (.5 * dist.flatten().sort().values[n_median].unsqueeze(dim=-1)) ** .5
-    return sigma
+    return sigma.log()
 
 
-class GaussianRBF(nn.Module):
+class KernelParameter:
     def __init__(
         self,
-        sigma: Optional[torch.Tensor] = None,
-        init_sigma_fn: Callable = sigma_median,
-        trainable: bool = False
+        value: torch.Tensor = None,
+        init_fn: Optional[Callable] = None,
+        requires_grad: bool = False,
+        requires_init: bool = False
+    ) -> None:
+        """
+        Parameter class for kernels.
+
+        Parameters
+        ----------
+        value
+            The pre-specified value of the parameter.
+        init_fn
+            The function used to initialize the parameter.
+        requires_grad
+            Whether the parameter requires gradient.
+        requires_init
+            Whether the parameter requires initialization.
+        """
+        super().__init__()
+        self.value = nn.Parameter(value if value is not None else torch.ones(1),
+                                  requires_grad=requires_grad)
+        self.init_fn = init_fn
+        self.requires_init = requires_init
+
+
+class BaseKernel(nn.Module):
+    def __init__(self, active_dims: list = None, feature_axis: int = -1) -> None:
+        """
+        The base class for all kernels.
+
+        Parameters
+        ----------
+        active_dims
+            Indices of the dimensions of the feature to be used for the kernel. If None, all dimensions are used.
+        feature_axis
+            Axis of the feature dimension.
+        """
+        super().__init__()
+        self.parameter_dict: dict = {}
+        if active_dims is not None:
+            self.active_dims = torch.as_tensor(active_dims)
+        else:
+            self.active_dims = None
+        self.feature_axis = feature_axis
+        self.init_required = False
+
+    @abstractmethod
+    def kernel_function(self, x: Union[np.ndarray, torch.Tensor], y: Union[np.ndarray, torch.Tensor],
```

Comment:
As we are in the `torch` specific submodule, I would actually fairly strongly argue that `np.ndarray` inputs should not be allowed. Note that the current implementation (e.g. line 125) would lead to device placement issues?

Reply:
Good pick. I initially followed the previous implementation and used extra device manipulation to bypass the issue (now removed). Will rewrite it to torch tensor only and double-check things.